ddl/index.go

```diff
@@ -532,138 +549,100 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
+		for i := 0; i < workerCnt; i++ {
 			wg.Add(1)
-			go d.doBackfillIndexTask(t, taskOpInfo, taskStartHandle, &wg)
-			doneHandle := <-taskOpInfo.nextCh
-			// There is no data to seek.
-			if doneHandle == taskStartHandle {
-				break
-			}
-			taskStartHandle = doneHandle + 1
+			workers[i].setTaskNewRange(baseHandle, baseHandle+taskBatch)
+			// TODO: Consider one worker to one goroutine.
+			go workers[i].doBackfillIndexTask(t, taskOpInfo, &wg)
+			baseHandle += taskBatch
 		}
 		wg.Wait()
 
-		retCnt := len(taskOpInfo.taskRetCh)
-		taskAddedCount, doneHandle, err := getCountAndHandle(taskOpInfo)
-		// Update the reorg handle that has been processed.
-		if taskAddedCount != 0 {
-			err1 := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
-				return errors.Trace(reorgInfo.UpdateHandle(txn, doneHandle+1))
-			})
-			if err1 != nil {
-				if err == nil {
-					err = err1
-				} else {
-					log.Warnf("[ddl] add index failed when update handle %d, err %v", doneHandle, err)
-				}
-			}
-		}
-
+		taskAddedCount, doneHandle, isEnd, err := getCountAndHandle(prevDoneHandle, workers)
 		addedCount += int64(taskAddedCount)
 		sub := time.Since(startTime).Seconds()
+		if err == nil {
+			err = d.isReorgRunnable()
+		}
 		if err != nil {
-			log.Warnf("[ddl] total added index for %d rows, this task add index for %d failed, take time %v",
-				addedCount, taskAddedCount, sub)
+			// Update the reorg handle that has been processed.
+			err1 := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
+				return errors.Trace(reorgInfo.UpdateHandle(txn, doneHandle))
+			})
+			log.Warnf("[ddl] total added index for %d rows, this task add index for %d failed, take time %v, update handle err %v",
+				addedCount, taskAddedCount, sub, err1)
 			return errors.Trace(err)
 		}
 		d.setReorgRowCount(addedCount)
 		batchHandleDataHistogram.WithLabelValues(batchAddIdx).Observe(sub)
 		log.Infof("[ddl] total added index for %d rows, this task added index for %d rows, take time %v",
 			addedCount, taskAddedCount, sub)
 
-		if retCnt < taskCnt {
+		prevDoneHandle, baseHandle = doneHandle, doneHandle+1
+		if isEnd {
 			return nil
 		}
 	}
 }
 
-// handleInfo records start and end handle that is used in a task.
-type handleInfo struct {
-	startHandle int64
-	endHandle   int64
-	isSent      bool // It ensures that the endHandle is assigned only once and is sent once.
-}
-
-func (h *handleInfo) isFinished(input int64) bool {
-	if !h.isSent || input < h.endHandle {
-		return false
-	}
-	return true
-}
-
-func getCountAndHandle(taskOpInfo *indexTaskOpInfo) (int64, int64, error) {
-	l := len(taskOpInfo.taskRetCh)
-	taskRets := make([]*taskResult, 0, l)
-	for i := 0; i < l; i++ {
-		taskRet := <-taskOpInfo.taskRetCh
-		taskRets = append(taskRets, taskRet)
-	}
-	sort.Sort(taskRetSlice(taskRets))
-
-	taskAddedCount, currHandle := int64(0), int64(0)
+func getCountAndHandle(prevDoneHandle int64, workers []*worker) (int64, int64, bool, error) {
```

Comment:
The variable `prevDoneHandle` can be removed. 
If the first worker failed, currHandle would be 0, then we don't need to update reorgInfo.

Reply:
We don't know when we need to update the reorgInfo's handle.

Reply:
Or we can just use the first worker's `startHandle`

Reply:
We need to know when seeking is end.